clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -689,6 +689,9 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
                     parsedRecords = new PartitionRecords<>(fetchOffset, tp, parsed);
                     ConsumerRecord<K, V> record = parsed.get(parsed.size() - 1);
                     this.sensors.recordsFetchLag.record(partition.highWatermark - record.offset());
+                } else if (partition.records.iterator().hasNext()) {
```

Comment:
This doesn't seem right. At a minimum, we're missing the '!' to invert the check. Additionally, iteration through a record set is not cheap (it may involve decompression), so we should try to use the work we've already done. In fact, maybe we could remove the check and compute the lag based on the high watermark and the current position of the consumer instead?